contrib/mypy/src/python/pants/contrib/mypy/tasks/mypy_task.py

```diff
@@ -144,19 +153,53 @@ def _collect_source_roots(self):
   def _interpreter_cache(self):
     return PythonInterpreterCache.global_instance()
 
-  def _run_mypy(self, py3_interpreter, mypy_args, **kwargs):
-    pex_info = PexInfo.default()
-    pex_info.entry_point = 'mypy'
+  def _get_mypy_pex(self, py3_interpreter: PythonInterpreter, *extra_pexes: PEX) -> PEX:
     mypy_version = self.get_options().mypy_version
-
-    mypy_requirement_pex = self.resolve_requirement_strings(
-      py3_interpreter, [f'mypy=={mypy_version}'])
-
-    path = os.path.realpath(os.path.join(self.workdir, str(py3_interpreter.identity), mypy_version))
-    if not os.path.isdir(path):
-      self.merge_pexes(path, pex_info, py3_interpreter, [mypy_requirement_pex])
-    pex = PEX(path, py3_interpreter)
-    return pex.run(mypy_args, **kwargs)
+    extras_hash = hash_utils.hash_all(hash_utils.hash_dir(Path(extra_pex.path()))
+                                      for extra_pex in extra_pexes)
+
+    path = Path(self.workdir,
+                str(py3_interpreter.identity),
+                f'{mypy_version}-{extras_hash}')
+    pex_dir = str(path)
+    if not path.is_dir():
```

Comment:
Style nit (feel free to ignore). Even though this leads to one extra line of code, I recommend inverting the conditional so that you can have less indentation.

```python
if path.is_dir()
  return PEX(pex_dir, py3_interpreter)
mypy_requirement_pex = ...
...
return PEX(pex_dir, py3_interpreter)
```

To me, makes a bit clearer the idea "if this already is created, simply return it; else, create it."

Reply:
This one feels small enough that I'm loathe to melt extra glaciers and babysit flaky shards. FWIW, I don't disagree, but I'd choose to just extract a function for the nested bit instead to keep the return DRY.

Reply:
Sgtm! Agreed also on extracting a function. But, yes, please merge :)

Reply:
Thanks for the review Eric.